src/Lucene.Net/Util/UnmanagedStringArray.cs

```diff
@@ -0,0 +1,185 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+
+namespace Lucene.Net.Util
+{
+    public unsafe class UnmanagedStringArray
+    {
+        private class Segment : IDisposable
+        {
+            public static long SegmentNumber;
+
+            public readonly int Size;
+            public readonly long Number;
+
+            public byte* Start;
+            public byte* CurrentPosition => Start + Used;
+            public int Free => Size - Used;
+            public int Used;
+
+            public Segment(int size)
+            {
+                Start = (byte*) Marshal.AllocHGlobal(size);
+                Used = 0;
+                Size = size;
+                Number = Interlocked.Increment(ref SegmentNumber);
+            }
+
+            public void Add(ushort size, out byte* position)
+            {
+                position = CurrentPosition;
+                *(ushort*) CurrentPosition = size;
+
+                Used += sizeof(short) + size;
+            }
+
+            public void Dispose()
+            {
+                GC.SuppressFinalize(this);
+                if (Start != null)
+                    Marshal.FreeHGlobal((IntPtr) Start);
+                Start = null;
+            }
+
+            ~Segment()
+            {
+                Dispose();
+            }
+        }
+
+        public struct UnmanagedString : IComparable
+        {
+            public byte* Start;
+            public long Owner;
+
+            public int Size => IsNull ? 0 : *(ushort*) Start;
+            public Span<byte> StringAsBytes => new Span<byte>(Start + sizeof(ushort), Size);
+            public bool IsNull => Start == default;
+            
+            public override string ToString()
+            {
+                return Encoding.UTF8.GetString(StringAsBytes.ToArray());
+            }
+
+            public static int CompareOrdinal(UnmanagedString strA, UnmanagedString strB)
+            {
+                if (strA.IsNull && strB.IsNull)
+                    return 0;
+
+                if (strB.IsNull)
+                    return 1;
+
+                if (strA.IsNull)
+                    return -1;
+
+                return strA.StringAsBytes.SequenceCompareTo(strB.StringAsBytes);
+            }
+
+            public static int CompareOrdinal(UnmanagedString strA, string strB)
+            {
+                if (strA.IsNull && strB == null)
+                    return 0;
+
+                if (strB == null)
+                    return 1;
+
+                if (strA.IsNull)
+                    return -1;
+
+                Span<byte> managed = Encoding.UTF8.GetBytes(strB);
```

Comment:
We shouldn't allocate a buffer here, instead, we want to use an already prepared buffer. Can use a thread local value, instead.